app/src/main/java/org/wikipedia/talk/NotificationDirectReplyHelper.kt

```diff
@@ -0,0 +1,118 @@
+                    Observable.just(Edit())
+                } else {
+                    var topicDepth = 0
+                    topic.replies?.lastOrNull()?.let {
+                        topicDepth = it.depth
+                    }
+                    val body = TalkTopicActivity.addDefaultFormatting(replyText, topicDepth)
+                    ServiceFactory.get(wiki).postEditSubmit(
+                        title.prefixedText, topic.id.toString(), null,
+                        DIRECT_REPLY_EDIT_COMMENT, if (AccountUtil.isLoggedIn) "user" else null, null, body,
+                        pair.second.revision, pair.first, null, null
+                    )
+                }
+            }
+            .subscribe({
+                if (it.edit?.editSucceeded == true) {
+                    waitForUpdatedRevision(context, wiki, title, it.edit.newRevId, notificationId)
+                } else {
+                    fallBackToTalkPage(context, title)
+                }
+            }, {
+                L.e(it)
+                fallBackToTalkPage(context, title)
+            })
+    }
+
+    private fun waitForUpdatedRevision(context: Context, wiki: WikiSite, title: PageTitle,
+                                       newRevision: Long, notificationId: Int) {
+        ServiceFactory.getRest(wiki).getTalkPage(title.prefixedText)
+            .delay(2, TimeUnit.SECONDS)
+            .subscribeOn(Schedulers.io())
+            .map { response ->
+                if (response.revision < newRevision) {
+                    throw IllegalStateException()
+                }
+                response.revision
+            }
+            .retry(20) { t ->
+                (t is IllegalStateException) || (t is HttpStatusException && t.code == 404)
+            }
+            .observeOn(AndroidSchedulers.mainThread())
+            .doOnTerminate {
+                cancelNotification(context, notificationId)
+            }
+            .subscribe({
+                // revisionForUndo = it
+                Toast.makeText(context, R.string.notifications_direct_reply_success, Toast.LENGTH_LONG).show()
+            }, { t ->
+                L.e(t)
+                fallBackToTalkPage(context, title)
+            })
+    }
+
+    private fun fallBackToTalkPage(context: Context, title: PageTitle) {
+        Toast.makeText(context, R.string.notifications_direct_reply_error, Toast.LENGTH_LONG).show()
+        context.startActivity(TalkTopicsActivity.newIntent(context, title, Constants.InvokeSource.NOTIFICATION))
+    }
+
+    private fun cancelNotification(context: Context, notificationId: Int) {
+        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.O) {
+            return
+        }
+        val notificationManager = context.getSystemService<NotificationManager>()!!
+        val notifications = notificationManager.activeNotifications!!
+        for (notification in notifications) {
+            if (notification.id == notificationId) {
```

Comment:
How about using: `notifications.find { it.id == notificationId }?.run { ... }`

or this for null-safe approach

`notificationManager?.activeNotifications?.find { it.id == notificationId }?.run {`

Reply:
correct!